packages/tools/devtools/devtools-core/src/ContainerRuntimeDevtools.ts

```diff
@@ -0,0 +1,165 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { TypedEventEmitter } from "@fluid-internal/client-utils";
+import type {
+	AttachState,
+	IAudience,
+	IContainerEvents,
+} from "@fluidframework/container-definitions/internal";
+import { ConnectionState } from "@fluidframework/container-loader";
+import type { IContainerRuntime } from "@fluidframework/container-runtime-definitions/internal";
+import type { IFluidLoadable } from "@fluidframework/core-interfaces";
+
+import { BaseDevtools } from "./BaseDevtools.js";
+import type { ContainerKey } from "./CommonInterfaces.js";
+import type { DecomposedContainer } from "./DecomposedContainer.js";
+import type { ContainerDevtoolsFeatureFlags } from "./Features.js";
+import type { InboundHandlers } from "./messaging/index.js";
+
+/**
+ * Properties for registering a container runtime with the Devtools.
+ * @alpha
+ */
+export interface ContainerRuntimeProps {
+	/**
+	 * The container runtime to register with the Devtools.
+	 *
+	 * @input
+	 */
+	readonly runtime: IContainerRuntime;
+
+	/**
+	 * Optional label for the container runtime replacing the guid generated by {@link @fluidframework/container-runtime#IContainerRuntime.generateDocumentUniqueId}.
+	 */
+	label?: string;
+}
+
+/**
+ * Properties for constructing a {@link ContainerRuntimeDevtools} instance.
+ */
+export interface ContainerRuntimeDevtoolsConstructorProps {
+	/**
+	 * The container key to use for this container runtime.
+	 */
+	containerKey: ContainerKey;
+
+	/**
+	 * The decomposed container wrapping the container runtime.
+	 */
+	container: DecomposedContainer;
+
+	/**
+	 * (optional) Data associated with the container runtime.
+	 */
+	containerData?: Record<string, IFluidLoadable>;
+}
+
+/**
+ * {@link IContainerDevtools} implementation for container runtimes.
+ *
+ * @remarks
+ *
+ * This class provides devtools functionality for container runtimes that don't support
+ * container-level operations like connect/disconnect/close. Container runtimes represent
+ * the internal runtime state and don't emit "closed" events like IContainer does.
+ *
+ * @sealed
+ */
+export class ContainerRuntimeDevtools extends BaseDevtools<DecomposedContainer> {
```

Comment:
It would probably be a good idea to add some unit tests for this type. We have some existing `ContainerDevtools` unit tests you can use for reference. They're not very exhaustive, but better than nothing :)